packages/crypto/src/crypto.ts

```diff
@@ -23,3 +23,45 @@ export const hashLeftRight = (input1: any, input2: any) =>
 export const hashOne = (input: any) => hash1([input])
 
 export { SNARK_FIELD_SIZE, genRandomSalt, stringifyBigInts, unstringifyBigInts }
+
+export const genEpochKey = (
+    identityNullifier: bigint,
+    attesterId: bigint | string,
+    epoch: bigint | number,
+    nonce: bigint | number,
+    maxEpochKey: bigint | number
+): bigint => {
+    const epochKey = hash4([
+        identityNullifier as any,
+        BigInt(attesterId),
+        epoch,
+        BigInt(nonce),
+    ]).valueOf()
+    // Adjust epoch key size according to epoch tree depth
+    const epochKeyModed = epochKey % BigInt(maxEpochKey)
+    return epochKeyModed
+}
```

Comment:
I think this function should be in `@unirep/core` because it is related to the protocol
the `@unirep/crypto` should be more general crypto functions

and it can be

```ts
export const genEpochKey = (
    identityNullifier: bigint,
    attesterId: bigint | string,
    epoch: bigint | number,
    nonce: bigint | number,
    maxEpochKey: bigint | number
): bigint => {
    const epochKey = hash4([
        identityNullifier,
        BigInt(attesterId),
        epoch,
        BigInt(nonce),
    ])
    // Adjust epoch key size according to epoch tree depth
    const epochKeyModed = epochKey % BigInt(maxEpochKey)
    return epochKeyModed
}
```

Reply:
It's nice having these functions in the crypto package because a client can calculate an epoch key or nullifier without needing to import core. We use these functions in the circuits/contracts packages so if we put `genEpochKey` in core we have a circular dependence.

The `@unirep/crypto` package is part of the unirep protocol I think, we're not writing crypto functions (e.g. implementing poseidon), these are just functions that are helpful when working with the protocol.

Reply:
`genEpochKey` is only used in test scripts in circuits and contracts
it is not a core function in circuits and contracts so it won't cause circular dependence

but it we put it in different package it misleads developers and `@unirep/crypto` should not handle what we **define** (e.g. epochKey, epochKeyNullifier) in the protocol

Reply:
I don't see how it misleads developers, the functions are named according to our protocol.

> @unirep/crypto should not handle what we define

I guess I don't understand this, why shouldn't we put our key calculation functions in our crypto package?

> it is not a core function in circuits and contracts so it won't cause circular dependence

ehhh these functions were defined in multiple places in the past. Maybe it's not a breaking circular dependence, but it's certainly a code smell imo

Reply:
Here is what we **define** in unirep protocol
https://docs.unirep.io/protocol/glossary

and hash, poseidon, SparseMerkleTree,... are not
and they are helpers of unirep protocol

yeah I agree they are defined in many places is annoying.
Or maybe we can rename it to `@unirep/utils` to handle these like protocol related functions but they are used in many packages

Reply:
Right I get that right now it's all completely generic structures, I don't see why it's bad to include unirep specific stuff in this package. It's under our namespace `@unirep` and it doesn't affect the other exports.

Reply:
but maybe `@unriep/crypto` is confusing to maintain all these functions.
So how do you think `@unirep/utils`?
I think it will affect like unirep social
Since it uses `genEpochKey()` functions
and previously it is in `@unriep/core`
https://github.com/Unirep/Unirep-Social/blob/40ff6ea26f514718409807210bd913655655c755/packages/frontend/src/context/User.ts#L7

Reply:
> but maybe @unriep/crypto is confusing to maintain all these functions.

I think it's okay as long as it's clear in the documentation.

> I think it will affect like unirep social

Yes there will be breaking changes for downstream applications. We can make it backward compatible by re-exporting it from `@unirep/core`.

Actually looking at [semver](https://semver.org/) maybe we should be calling this v2 because it has other breaking changes.

Reply:
>  So how do you think @unirep/utils?

I really don't like this. What are utils? Hash functions are utils, making an epoch key is too, maybe a tree structure is a util.

Using `utils` anywhere is a code smell to me

Reply:
> Yes there will be breaking changes for downstream applications. We can make it backward compatible by re-exporting it from @unirep/core.

I don't think in the breaking changes we don't really care about backward compatible. We will anyway change unirep social according to `v1.1`

> Actually looking at [semver](https://semver.org/) maybe we should be calling this v2 because it has other breaking changes.

I think it is okay to call it `v2`

> I really don't like this. What are utils? Hash functions are utils, making an epoch key is too, maybe a tree structure is a util.

yes they are
I think it will be better than put it all in `@unirep/crypto` since some of them are not crypto related
or hash function in `@unirep/crypto` and another package `@unirep/utils`

Reply:
> I think it is okay to call it v2

👍

>  I think it will be better than put it all in @unirep/crypto since some of them are not crypto related

But something like `genEpochKey` _is_ crypto related. It calculates a specific hash using some inputs. It's like calculating a public key in tls or ssh or semaphore. We put it in the crypto package because it's part of the unirep cypto-protocol.

We could technically break these 4 functions into a separate package, but is that worth the complexity? Both in terms of managing packages, and users figuring out what the different packages are?

Reply:
For me, `genEpochKey` is not a generally known crypto related function (like `SparseMerkleTree` or `hash`). It is like a specific function used in unirep. 

> We could technically break these 4 functions into a separate package, but is that worth the complexity? Both in terms of managing packages, and users figuring out what the different packages are?

Yeah so I suggest them to be in `@unirep/utils` to be easily found them all
and they don't have to be crypto related

Reply:
like `ethers` also put hash in `utils`
https://docs.ethers.io/v5/api/utils/hashing/

Reply:
But notice that ethers doesn't have a `utils` _and_ a `crypto` package. I'd be fine with changing the entire `crypto` package to be `utils`. I just don't want to maintain many packages that have few functions.

Reply:
I agree with that the entire `crypto` to be `utils` 👍🏻

Reply:
ok, documented in #166